### PR TITLE
Some attempts at cross-platform compatibility

### PR DESCRIPTION
**Pull request by N00byKing**
Hi!
I've been trying to get the official launcher to work on linux, with little success.
However, after finding this one it turns out that even if linux may not be officially supported, it does at least compile!
Sadly pretty much nothing else worked.

This PR replaces some calls with cross-platform alternatives; enough for the launcher to support all but launching the game on Linux.
It should have (little to) no effect on Windows users.

### CODE REVIEW

File: src/Program.cs [hunk 1]
```diff
@@ -121,10 +116,7 @@ protected override void OnStartupNextInstance(int processId, string[] args)
                         }
                         if (!mainWindow.IsActive)
                         {
-                            if (!Leayal.Shared.Windows.UnmanagedWindowsHelper.SetForegroundWindow(mainWindow))
-                            {
-                                mainWindow.Activate();
-                            }
+                            mainWindow.Activate();
```
Comment (N00byKing) [inline src/Program.cs:119]: This is the only place where functionality may be impacted. If needed, I can change it to only use `SetForegroundWindow` on Windows

File: src/Windows/MainWindow.Helpers.cs [hunk 2]
```diff
@@ -54,12 +55,15 @@ private static void AdjustManualSize(MsBox.Avalonia.Dto.MessageBoxStandardParams
         {
             if (sizeToContent == SizeToContent.Manual)
             {
-                int screenWidth = PInvoke.GetSystemMetrics(global::Windows.Win32.UI.WindowsAndMessaging.SYSTEM_METRICS_INDEX.SM_CXVIRTUALSCREEN),
-                    screenHeight = PInvoke.GetSystemMetrics(global::Windows.Win32.UI.WindowsAndMessaging.SYSTEM_METRICS_INDEX.SM_CYVIRTUALSCREEN);
-                msgboxparams.MinWidth = 300;
-                msgboxparams.MinHeight = 200;
-                msgboxparams.Width = screenWidth / 2;
-                msgboxparams.Height = screenHeight / 2;
+            int screenWidth = 1920, screenHeight = 1080;
```
Comment (N00byKing) [inline src/Windows/MainWindow.Helpers.cs:58]: Currently, no cross-platform way exists to get the screen dimensions that I am aware of.
This implements a fallback to the most common resolution used (via Steam HW Survey), but will use actual size on Windows systems.

**Comment by Leayal** [comment]
For getting screen size, you can try create process `xrandr` and get its output from the pipe, after that, parse the text to get the screen's size. But this would impact performance if we were to do this on every dialog showup, so get the screen once at startup and use the cached size value, or some kind of time-limited caching to avoid wrong cache result when screen size changes.

Psuedo sample, may need some corrects for it to actually work:
```csharp
public bool TryGetDisplaySize(out int width, out int height)
{
    System.Diagnostics.Process proc = new System.Diagnostics.Process("xrandr");
    p.StartInfo.UseShellExecute = false;
    p.StartInfo.RedirectStandardOutput = true;
    p.Start();
    p.WaitForExit();
    string output = p.StandardOutput.ReadToEnd();
    var match = System.Text.RegularExpressions.Regex.Match(output, @"(\d+)x(\d+)\+0\+0");
    if (match == null) return false;
    var w = match.Groups[1].Value;
    var h = match.Groups[2].Value;
    if (!int.TryParse(w, out width)) return false;
    if (!int.TryParse(h, out height)) return false;
}
```
P/s: I have never programmed Linux Desktop apps so I'm pretty much know nothing about this. I'll look into this, though.
Can I ask how do you play the game on Linux? Using Wine, Proton or something?

**Comment by N00byKing** [comment]
`xrandr` might work for now, but depends on having a X server running, which is also not as portable as one would like... Might make for a good stopgap though.

I am using Wine (Translate most Windows syscalls to linux calls) to run the game, combined with DXVK (Translate DirectX to Vulkan) :)
Essentially, I use the launcher (without wine or dxvk, as I can compile it to native binary) to download the game / verify files / update, then when its done close it and run the game manually: `wine start game/Game/Binaries/Win64/Game.exe -FeatureLevelES31 -ChannelID=seasun`
Works like a charm once I get past actually downloading the game, which was the hard part xD
But due to this project significantly easier, so thanks for making the launcher!

**Comment by Leayal** [comment]
For platform-specific code, you can use something like...
```csharp
int screenWidth, screenHeight;
if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
{
    screenWidth = PInvoke.GetSystemMetrics(global::Windows.Win32.UI.WindowsAndMessaging.SYSTEM_METRICS_INDEX.SM_CXVIRTUALSCREEN);
    screenHeight = PInvoke.GetSystemMetrics(global::Windows.Win32.UI.WindowsAndMessaging.SYSTEM_METRICS_INDEX.SM_CYVIRTUALSCREEN);
}
else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
{
    TryGetDisplaySize(out screenWidth, out screenHeight);
}
else
{
    throw new NotSupportedException();
}
```

**Comment by Leayal** [comment]
> I am using Wine (Translate most Windows syscalls to linux calls) to run the game, combined with DXVK (Translate DirectX to Vulkan)

Hm...Can you try running this launcher on Wine, too?

**Comment by N00byKing** [comment]
.NET apps run pretty bad in wine from my experience, sadly. This requires installing .NET Runtime inside Wine, which often breaks. Compiling it for native use is the most stable approach by far

**Comment by Leayal** [comment]
> This requires installing .NET Runtime inside Wine,

The launcher doesn't requires this, it already has the .NET Runime within the binary (therefore, the .exe file has big size...). It's a standalone app, so you just need to run it directly inside Wine without any additional software installations.

**Comment by Leayal** [comment]
> run the game manually: `wine start game/Game/Binaries/Win64/Game.exe -FeatureLevelES31 -ChannelID=seasun`

I'll also add this to the launcher, to run the game on Linux via Wine. But it will be a quite while later...
I'm currently at work so no coding for me.

**Comment by N00byKing** [comment]
I tried it out now, it starts but does not work correctly:
- Clicking Events does not work, but for some reason Banners do
- Settings Icon is broken
- Additional Windows open slower, and text in new windows does not render correctly (Spacing between letters varies a lot)
- Launcher crashed after selecting game location

These issues do not appear when compiling natively (as long as the patches from the PR are applied)

> > run the game manually: `wine start game/Game/Binaries/Win64/Game.exe -FeatureLevelES31 -ChannelID=seasun`
> 
> I'll also add this to the launcher, to run the game on Linux via Wine. But it will be a quite while later... I'm currently at work so no coding for me.

Thank you for considering adding this! However, the way the launcher detects if the game is running might be hard to port; it uses the most Windows-native functions by far.

**Comment by Leayal** [comment]
> Thank you for considering adding this! However, the way the launcher detects if the game is running might be hard to port; it uses the most Windows-native functions by far.

I intended to change this feature to `Only detect if game is running when pressing Game Start button`, as well as find a way to make it work on Linux, albeit no promises. I still feel like wasting CPU cycles to poll processes is a bad idea, I originally wanted this to prevent updating the game while playing, but thinking about it now, it's really uselessly unnecessary to do it this way.

**Comment by Leayal** [comment]
May I ask you which Linux distro you're using?

**Comment by Leayal** [comment]
I have put some changes for Linux, but I can't test it since I don't even know which Linux distro to test....
As well as limited time (since I only maintain this launcher as a hobby).

Please check it out in the [`dev-linux` branch](https://github.com/Leayal/SnowBreakLauncher-Dotnet/tree/dev-linux), if it can compile natively for Linux and can be run without issue? If it does have issue, please let me know, @N00byKing

**Comment by Leayal** [comment]
@N00byKing 
Please check out the new release which is an attempt to support Linux OS: https://github.com/Leayal/SnowBreakLauncher-Dotnet/releases/tag/1.1.1

If you have any issues, please let me know. I'll close this in the after 7-day without any further comments and consider this is completed.

**Comment by N00byKing** [comment]
Sorry, my friend is moving and I have had literally no time to do anything...
I will try to test it in the next couple of days

**Comment by Leayal** [comment]
No rush. What I meant was just close this if no comments.
You can still create a new issue and report the problem in case this pull request was closed.

**Comment by Wildikdog** [comment]
I'm not the person asked but also noticed the update to try to support Linux. So far the only real problems I can see is that I can't seemingly update or repair game files as the manifest.json is in use with another process but I can't tell if that's because I just linked the launcher to my already downloaded version, or if it's because of my bungled attempts to get Dependent working (Because I'm bad at this stuff probably nothing wrong with the instructions themselves), and the inability to choose Wine, or DXVK versions which I assume is because the launcher just uses the system's wine. 

Regardless, thankful to see an attempt even if it never gets all the features, it'd be nice to be able to update on Linux for me.

EDIT: Realized I never showed the error code
https://pastebin.com/Tr1a5WaB

**Comment by Leayal** [comment]
> and the inability to choose Wine, or DXVK versions which I assume is because the launcher just uses the system's wine.

Since I'm not well-versed in Linux, I thought when installing Wine using a package manager of the distro, `wine` will be added to `PATH` environment variable of the system, which then when running `wine` without "pathlessly" (running the filename without telling where it is), the system will look it up by itself for you. But I admit that if wine is built from source or not "installed" by usual way, it's busted. I can add wine's path selecting, or custom launch arguments for wine.

...In fact, I don't even know how to select or customize Wine (like selecting DXVK versions or things like that).

> Realized I never showed the error code
> https://pastebin.com/Tr1a5WaB

I can't access pastebin, though you can just post the file content to [Github Gist](https://gist.github.com/), this thing is somewhat very similar to Pastebin, but it's on Github.

Can I also ask for your Linux distro's name and version? There are so many and with different graphic system that I don't know where I should suppose to swim toward in the vast sea of distro...

**Comment by Leayal** [comment]
> Realized I never showed the error code
> https://pastebin.com/Tr1a5WaB

I can't access pastebin, though you can just post the file content to [Github Gist](https://gist.github.com/), this thing is somewhat very similar to Pastebin, but it's on Github.

**Comment by Wildikdog** [comment]
> Since I'm not well-versed in Linux, I thought when installing Wine using a package manager of the distro, `wine` will be added to `PATH` environment variable of the system, which then when running `wine` without "pathlessly" (running the filename without telling where it is), the system will look it up by itself for you. But I admit that if wine is built from source or not "installed" by usual way, it's busted. I can add wine's path selecting, or custom launch arguments for wine.
> 
Oh no it's not busted or anything like that from my point of view. I was more meaning to say that the ability to change versions and such is a nicety in Linux focused launchers but works just fine, especially as this is a side project and Linux support isn't even the focus.

> 
> I can't access pastebin, though you can just post the file content to [Github Gist](https://gist.github.com/), this thing is somewhat very similar to Pastebin, but it's on Github.
> 
https://gist.github.com/Wildikdog/559241ee6f553f7cbb5e08a475768206 
I think that's the correct way to do it, sorry never used either pastebin or Github Gist until now.

> Can I also ask for your Linux distro's name and version? There are so many and with different graphic system that I don't know where I should suppose to swim toward in the vast sea of distro...
>
I use Nobara 38, which is a offshoot of Fedora, admittedly I don't know much about programming, much less programming for Linux applications but I don't think you should need to force yourself to program towards any one distro in particular, as long as Dependencies are listed out I wouldn't think there is any issue getting it up and running on any distro, user error aside.

**Comment by Leayal** [comment]
> but I don't think you should need to force yourself to program towards any one distro in particular

I don't want that, either. That's one way to hell, thinking each distro needs each different code path.

What I want to know was the eco system that you're using. For example, for Linux, there are different desktop environments (KDE, Gnome, Xfce, etc...). While it's Avalonia helped me dealing with this variations, if I still need to invoke the specific desktop's API, I will have to know their function names for invocation.

**Comment by Wildikdog** [comment]
> I don't want that, either. That's one way to hell, thinking each distro needs each different code path.
> 
Meant more like designing around say Ubuntu or another popular Distro, and just hoping for the best for the rest, but that also sounds like hell.
> What I want to know was the eco system that you're using. For example, for Linux, there are different desktop environments (KDE, Gnome, Xfce, etc...). While it's Avalonia helped me dealing with this variations, if I still need to invoke the specific desktop's API, I will have to know their function names for invocation.
>
Ah, is that what you meant that when you said version? I use Gnome, not sure on what other info you'd need.

**Comment by Leayal** [comment]
I still don't know what the cause of `the manifest.json is in use with another process`.
If it still happens, can you check what processes are using the file (aside from the launcher)?

**Comment by Wildikdog** [comment]
> I still don't know what the cause of `the manifest.json is in use with another process`. If it still happens, can you check what processes are using the file (aside from the launcher)?

I checked using both fuser and lsof to double check but both say only the Launcher itself is using manifest.json, checked just to see if it's some potential issue on my end and my Steam Deck says the same thing.

**Comment by Leayal** [comment]
@Wildikdog Can you try this build (link below)?
https://github.com/Leayal/SnowBreakLauncher-Dotnet/releases/tag/1.1.2-pre

Please let me know if the issue is still occuring or not.
In case it still occurs, please check if the launcher is running multiple instances or not.

**Comment by Wildikdog** [comment]
File integrity check seems to have finished correctly now on my PC and Deck, and game was still bootable after so no problems there, it seems, will check if there are any issues with patching when one comes out. Just to make sure, in case there was any user error involved, when linking to an already existing install the file manager was looking for a "Game Client File", the following dialog box mentioned Manifest.json so I assumed that was the Game Client File the manager wanted me to link to, is that correct?

Also tried a fresh game install and it errored out at the end
https://gist.github.com/Wildikdog/9d5ee4c9881ac50b3952fc90c110f99c

**Comment by Leayal** [comment]
@Wildikdog Another build which should resolve that `A positive capacity must be specified for a Memory Mapped File backed by an empty file.` issue.
https://github.com/Leayal/SnowBreakLauncher-Dotnet/releases/tag/1.1.3-pre

**Comment by Wildikdog** [comment]
@Leayal Download finished correctly, problem does seem to be fixed.

**Comment by Leayal** [comment]
> problem does seem to be fixed.

~~One less loose end.~~.

About Wine customizing, can I ask you about how to? For example, selecting DXVK version or things like that?

**Comment by Leayal** [comment]
I'll close this pull request as the launcher seems to be stable enough on Linux.
Or at least that I'm unaware of the bugs existing in the launcher's code path for Linux.

The release for Linux and Windows can be downloaded here: https://github.com/Leayal/SnowBreakLauncher-Dotnet/releases/tag/1.2.0

Please use this version. And, if there's a problem, you can [create a new issue here](https://github.com/Leayal/SnowBreakLauncher-Dotnet/issues).

Best regard and have fun with the game, @Wildikdog , @N00byKing

**Pull request closed without merging**